lib/Service/ImageService.php

```diff
@@ -0,0 +1,676 @@
+		} catch (ServerException | ClientException $e) {
+			//$response = $e->getResponse();
+			//if ($response->getStatusCode() === 401) {
+			$this->logger->warning('Impossible to download image', ['exception' => $e]);
+			return ['error' => 'Impossible to download image'];
+		} catch (ConnectException $e) {
+			$this->logger->error('Connection error', ['exception' => $e]);
+			return ['error' => 'Connection error'];
+		} catch (Throwable | Exception $e) {
+			$this->logger->error('Unknown download error', ['exception' => $e]);
+			return ['error' => 'Unknown download error'];
+		}
+	}
+
+	/**
+	 * Actually delete attachment files which are not pointed in the markdown content
+	 *
+	 * @param int $fileId
+	 * @return int The number of deleted files
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 * @throws \OCP\Lock\LockedException
+	 */
+	public function cleanupAttachments(int $fileId): int {
+		$textFile = $this->rootFolder->getById($fileId);
+		if (count($textFile) > 0 && $textFile[0] instanceof File) {
+			$textFile = $textFile[0];
+			if ($textFile->getMimeType() === 'text/markdown') {
+				// get IDs of the files inside the attachment dir
+				$attachmentDir = $this->getOrCreateAttachmentDirectoryForFile($textFile);
+				$attachmentsByName = [];
+				foreach ($attachmentDir->getDirectoryListing() as $attNode) {
+					$attachmentsByName[$attNode->getName()] = $attNode;
+				}
+
+				$contentAttachmentNames = $this->getAttachmentNamesFromContent($textFile->getContent());
+
+				$toDelete = array_diff(array_keys($attachmentsByName), $contentAttachmentNames);
+				foreach ($toDelete as $name) {
+					$attachmentsByName[$name]->delete();
+				}
+				return count($toDelete);
+			}
+		}
+		return 0;
+	}
+
+
+	/**
+	 * Get attachment file names listed in the markdown file content
+	 *
+	 * @param string $content
+	 * @return array
+	 */
+	private function getAttachmentNamesFromContent(string $content): array {
+		$matches = [];
+		preg_match_all(
+			'/\!\[[^\[\]]+\]\(text:\/\/image\?[^)]*imageFileName=([^)&]+)\)/',
+			$content,
+			$matches,
+			PREG_SET_ORDER
+		);
+		return array_map(function (array $match) {
+			return $match[1] ?? null;
+		}, $matches);
+	}
+
+	/**
+	 * @param File $source
+	 * @param File $target
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 * @throws \OCP\Lock\LockedException
+	 */
+	public function moveAttachments(File $source, File $target): void {
+		// if the parent directory has changed
+		if ($source->getParent()->getPath() !== $target->getParent()->getPath()) {
+			$sourceAttachmentDir = $this->getAttachmentDirectoryForFile($source);
+			// if there is an attachment dir for this file
+			// and it is in the same directory as the source file
+			if ($sourceAttachmentDir !== null
+				&& $source->getParent()->getId() === $sourceAttachmentDir->getParent()->getId()
```

Comment:
I improved the comment. This is tricky.
* move:
    * if done by the .md file's owner, the attachment dir is next to the .md file => we should move the attachment dir
    * if someone else having access does it, as the file won't move in the owner's storage => we don't move the attachment dir
* copy is different because we duplicate the attachments, this is done regardless if the file is copied by its owner or someone else.

Does it make sense? :grin:

Reply:
Ah, makes sense, thanks for the explanation! Maybe just use the last comment?

```php
		// Move attachement dir only if .md file is owned by it's owner.
```